src/cs3iface.py

```diff
@@ -9,137 +9,215 @@
-    # write the file. In a future implementation, we should find a way to only update the required chunks...
-    written = f.write(content)
-    f.close()
-    if written != size:
-      raise IOError('Written %d bytes but content is %d bytes' % (written, size))
-  except OSError as e:
-    log.warning('msg="Error writing to file" filename="%s" error="%s"' % (filename, e))
-    raise IOError(e)
-
-def renamefile(_endpoint, origfilename, newfilename, ruid, rgid):
-  '''Rename a file from origfilename to newfilename on behalf of the given uid, gid.'''
-  try:
-    os.rename(_getfilename(origfilename), _getfilename(newfilename))
-  except (FileNotFoundError, PermissionError) as e:
-    raise IOError(e)
-
-def removefile(_endpoint, filename, _ruid, _rgid, _force=0):
-  '''Remove a file on behalf of the given uid, gid.
-     The force argument is irrelevant and ignored for local storage.'''
+    fileget = requests.get(url=filedownloadres.download_endpoint,
+                           headers={'x-access-token': _authenticate(userid)})
+  except requests.exceptions.RequestException as e:
+    ctx['log'].error('msg="Exception when downloading file from Reva" reason="%s"' % e)
+    yield IOError(e)
+  tend = time.clock()
+  data = fileget.content
+  if fileget.status_code != http.client.OK:
+    ctx['log'].error('msg="Error downloading file from Reva" code="%d" reason="%s"' % (fileget.status_code, fileget.reason))
+    yield IOError(fileget.reason)
+  else:
+    ctx['log'].info('msg="File open for read" filepath="%s" elapsedTimems="%.1f"' % (filepath, (tend-tstart)*1000))
+    for i in range(0, len(data), ctx['chunksize']):
+      yield data[i:i+ctx['chunksize']]
+
+
+def writefile(_endpoint, filepath, userid, content, _noversion=0):
+  '''Write a file using the given userid as access token. The entire content is written
+    and any pre-existing file is deleted (or moved to the previous version if supported).
+    The noversion flag is currently not supported.'''
+  tstart = time.clock()
+  # prepare endpoint
+  req = sp.InitiateFileUploadRequest(ref=spr.Reference(path=filepath))
+  fileuploadres = ctx['cs3stub'].InitiateFileUpload(
+      request=req, metadata=[('x-access-token', _authenticate(userid))])
+  if fileuploadres.status.code != cs3code.CODE_OK:
+    ctx['log'].debug('msg="Failed to initiateFileUpload on write" filepath="%s" reason="%s"' % \
+                     (filepath, fileuploadres.status.message))
+    raise IOError(fileuploadres.status.message)
+  ctx['log'].debug('msg="writefile: InitiateFileUploadRes returned" endpoint="%s"' % fileuploadres.upload_endpoint)
+
+  # Upload
   try:
-    os.remove(_getfilename(filename))
-  except (FileNotFoundError, PermissionError, IsADirectoryError) as e:
+    # TODO: Use tus client instead of PUT
+    headers = {
+        'Tus-Resumable': '1.0.0',
+        'x-access-token':  _authenticate(userid)
+    }
+    fileput = requests.put(url=fileuploadres.upload_endpoint, data=content, headers=headers)
```

Comment:
I think "fileput" is a bit strange name, how about `uploadres` or something similar?

Reply:
then let's call this `putres`, as upload is more the CS3 API terminology